src/app/file-tree/file-tree.component.ts

```diff
@@ -0,0 +1,105 @@
+  absolutePath: string;
+  expandable: boolean;
+}
+
+/**
+ * TODO: Shift the file (not folders) to the left so that it's aligned with the folders
+ * TODO: Title? Dialog actions?
+ * TODO: In the tree, it's not obvious to click on file to select
+ * TODO: When to actually use this instead of the simpler dropdown? Over a certain amount of files?
+ * TODO: What's my current selected file?
+ * Stretch TODO: Make sure there's always more than one child node, otherwise collapse child with parent (i.e. instead of parentname => childname, it's just parentname/childname)
+ */
+@Component({
+  selector: 'app-file-tree',
+  templateUrl: './file-tree.component.html',
+})
+export class FileTreeComponent {
+  /** The TreeControl controls the expand/collapse state of tree nodes.  */
+  treeControl: FlatTreeControl<FlatTreeNode>;
+
+  /** The TreeFlattener is used to generate the flat list of items from hierarchical data. */
+  treeFlattener: MatTreeFlattener<FileNode, FlatTreeNode>;
+
+  /** The MatTreeFlatDataSource connects the control and flattener to provide data. */
+  dataSource: MatTreeFlatDataSource<FileNode, FlatTreeNode>;
+
+  constructor(private matDialogRef: MatDialogRef<FileTreeComponent>, @Inject(MAT_DIALOG_DATA) public data: SourceFile[]) {
+    this.treeFlattener = new MatTreeFlattener(this.transformer, this.getLevel, this.isExpandable, this.getChildren);
+
+    this.treeControl = new FlatTreeControl(this.getLevel, this.isExpandable);
+    this.dataSource = new MatTreeFlatDataSource(this.treeControl, this.treeFlattener);
+    this.dataSource.data = this.convertSourceFilesToTree(data);
+  }
+
+  /** Transform the data to something the tree can read. */
+  transformer(node: FileNode, level: number): FlatTreeNode {
+    return {
+      name: node.name,
+      level,
+      absolutePath: node.absolutePath,
+      expandable: !!node.children && node.children.length > 0,
+    };
+  }
+
+  /** Get the level of the node */
+  getLevel(node: FlatTreeNode): number {
+    return node.level;
+  }
+
+  /** Get whether the node is expanded or not. */
+  isExpandable(node: FlatTreeNode): boolean {
+    return node.expandable;
+  }
+
+  /** Get whether the node has children or not. */
+  hasChild(index: number, node: FlatTreeNode): boolean {
+    return node.expandable;
+  }
+
+  /** Get the children for the node. */
+  getChildren(node: FileNode): FileNode[] | null | undefined {
+    return node.children;
+  }
+
+  selectFile(node: FileNode) {
+    this.matDialogRef.close(node.absolutePath);
+  }
+
+  private convertSourceFilesToTree(sourceFiles: SourceFile[]): FileNode[] {
+    // Everything is assumed to be an absolute path starting with "/", removing that slash as it's not needed for path.split
+    const paths = sourceFiles.map((sourceFile) => sourceFile.absolutePath.substring(1));
+    let result: FileNode[] = [];
+    let level = { result };
+    paths.forEach((path) => {
+      path.split('/').reduce((accumulator, filename, i, a) => {
```

Comment:
Suggest better names for `i` and `a` -- I hadn't used reduce in a while and had to go look up what they are.

Also confused by `filename` -- it's a path segment.

Reply:
it's actually a filename or folder name. the i and a are named that way because they weren't really used so it doesn't matter